src/Functions/map.cpp

```diff
@@ -239,6 +246,207 @@ class FunctionMapFromArrays : public IFunction
+
+    bool isSuitableForShortCircuitArgumentsExecution(const DataTypesWithConstInfo & /*arguments*/) const override { return true; }
+
+    DataTypePtr getReturnTypeImpl(const ColumnsWithTypeAndName & arguments) const override
+    {
+        FunctionArgumentDescriptors args{
+            {"str_expr", &isStringOrFixedString<IDataType>, nullptr, "String or FixedString"}
+        };
+        FunctionArgumentDescriptors optional_args{
+            {"pair_delim", &isString<IDataType>, isColumnConst, "const String"},
+            {"key_value_delim", &isString<IDataType>, isColumnConst, "const String"}
+        };
+
+        validateFunctionArgumentTypes(*this, arguments, args, optional_args);
+
+        DataTypePtr string_type = std::make_shared<DataTypeString>();
+        DataTypePtr string_type_nullable = std::make_shared<DataTypeNullable>(string_type);
+        DataTypes key_value_type{string_type, string_type_nullable};
+
+        return std::make_shared<DataTypeMap>(key_value_type);
+    }
+
+    bool useDefaultImplementationForConstants() const override { return true; }
+
+    static void executeString(const ColumnString * col_str_in, ColumnPtr & col_res, size_t input_rows_count, char pair_delim, char key_value_delim)
+    {
+        String pair_delim_str(1, pair_delim);
+
+        const ColumnString::Chars & in_vec = col_str_in->getChars();
+        const ColumnString::Offsets & in_offsets = col_str_in->getOffsets();
+        size_t size = in_offsets.size();
+
+        MutableColumnPtr keys_data = DataTypeString().createColumn();
+        MutableColumnPtr values_data = DataTypeString().createColumn();
+        MutableColumnPtr offsets = DataTypeNumber<IColumn::Offset>().createColumn();
+        ColumnUInt8::MutablePtr values_null_map = ColumnUInt8::create();
+
+        keys_data->reserve(input_rows_count);
+        values_data->reserve(input_rows_count);
+        offsets->reserve(input_rows_count);
+        values_null_map->reserve(input_rows_count);
+
+        ColumnString::Offset prev_res_offset = 0, prev_in_offset = 0;
+
+        std::vector<String> key_value_pairs;
+
+        for (size_t i = 0; i < size; ++i)
+        {
+            key_value_pairs.clear();
+            const char * pos_in = reinterpret_cast<const char *>(&in_vec[prev_in_offset]);
+            size_t curr_size = in_offsets[i] - prev_in_offset - 1;
+            // splitting the string into key-value pairs
+            boost::split(key_value_pairs, std::string_view(pos_in, curr_size), boost::is_any_of(pair_delim_str));
+            // preparing the key, value columns for the map
+            for (size_t j = 0; j < key_value_pairs.size(); ++j)
+            {
+                const std::string_view str_view = key_value_pairs[j];
+                size_t delim_pos = str_view.find(key_value_delim);
+                if (delim_pos != std::string_view::npos)
+                {
+                    keys_data->insert(str_view.substr(0, delim_pos));
```

Comment:
If I am not mistaken, this constructs a temporary `Field` object, which is costly since it has to copy a string (potentially allocating memory for that).

You may want to consider using `Colunm::insertData` instead or use some specialized helper, like `ColumnStringHelpers::WriteHelper`